service/src/main/java/bio/terra/externalcreds/services/TokenProviderService.java

```diff
@@ -0,0 +1,82 @@
+package bio.terra.externalcreds.services;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.externalcreds.auditLogging.AuditLogEvent;
+import bio.terra.externalcreds.auditLogging.AuditLogEventType;
+import bio.terra.externalcreds.auditLogging.AuditLogger;
+import bio.terra.externalcreds.config.ExternalCredsConfig;
+import bio.terra.externalcreds.models.LinkedAccount;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.HashSet;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.oauth2.core.OAuth2AuthorizationException;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class TokenProviderService extends ProviderService {
+
+  public TokenProviderService(
+      ExternalCredsConfig externalCredsConfig,
+      ProviderClientCache providerClientCache,
+      OAuth2Service oAuth2Service,
+      LinkedAccountService linkedAccountService,
+      AuditLogger auditLogger,
+      ObjectMapper objectMapper) {
+    super(
+        externalCredsConfig,
+        providerClientCache,
+        oAuth2Service,
+        linkedAccountService,
+        auditLogger,
+        objectMapper);
+  }
+
+  public Optional<LinkedAccount> createLink(
+      String providerName,
```

Comment:
This should be a `Provider`, not a `String`. That way, we can do validation on it.

Reply:
I'll change it to be consistent with https://github.com/DataBiosphere/terra-external-credentials-manager/pull/168. I'm not sure that we need to check if this is a valid provider here since that already happens upstream. Unless you mean validating that it is a token-based provider?